### ISSUE TITLE
How to wait on a group of tasks with Huey

### ISSUE THREAD

**Issue by distefam** [coleifer/huey#286]
I'd like to migrate from Celery to Huey but there is one potential sticking point. With Celery, I have a task that spawns a bunch of parallel tasks (a [group](http://docs.celeryproject.org/en/latest/userguide/canvas.html#groups)) and waits on all of those parallel tasks to finish before executing a callback using a [chord](http://docs.celeryproject.org/en/latest/userguide/canvas.html#chords). For context, the parallel tasks fetch something from an external API, then the callback batch inserts this data into my Postgres database. 

I'm trying to think of how to accomplish something similar with Huey and came up with the following:
1. Just have each task execute an API call and insert the result into the database individually. This is the simplest but potentially hammers the database if a bunch of API calls return simultaneously.
2. Have a master task that spawns a bunch of API call tasks as well as another task that checks the result of all those tasks with a backoff delay.

I'm not really satisfied with either of these approaches and was hoping someone might have a better suggestion.

**Comment by coleifer** [comment]
I just read the celery docs for `group` and `chord`...my own opinion is that it's hideously complicated and hard to understand (littered with `.s()`, `.si()`, pipes, etc). I recognize the engineering talent that's required to make a library like celery, but if I was reviewing code that looked like those examples I'd definitely take issue with it on account of it being almost unreadable.

That being said, I think there's probably an analogous pattern in threading for spawning a bunch of threads (which may put values in a queue), then waiting for all to join and doing something with the values put in the queue.

To implement that kind of thing with huey you'll want to do something like:

```python
@huey.task()
def fetch_from_api(resource):
    return api.get(resource)

@huey.task()
def initialize_data_from_api():
    resources = ['foo', 'bar', 'baz', 'something']

    # Spawn 4 tasks.
    tasks = [fetch_from_api(resource) for resource in resources]

    # Block this task until the 4 fetch tasks have returned.
    results = [task_result.get() for task_result in tasks]

    # Insert the resulting data into the db.
    insert_into_db(results)
```

In the above code you have a master task that spawns individual API-call tasks and then *blocks* until all tasks have been executed and returned, before inserting the resulting data into the db. Obviously for the above to work you'll need at least 2 workers, since the master task is going to block up a worker for the duration of the API calls.

Is there any particular reason you're migrating away from celery?

**Comment by coleifer** [comment]
Just added support for pipelining. No docs yet, but might interest you:

312370f48f7f0e61a5e455cb41f135a4c37115f5

**Comment by coleifer** [comment]
It's not the same as a chord, where you've got a group running in parallel and a callback when they're all done, but just FYI.

**Comment by distefam** [comment]
Thank you so much for the quick and detailed reply. I wanted to try it out before responding but haven't had a chance to yet. I'll be working on it today though and I'll also be sure to check out the new pipelining features.

You pretty much nailed my primary reason for wanting to move away from Celery. As you noted, the documentation and examples are frustratingly complex and in my opinion insufficient (just one example: chording, it took me far longer than necessary to realize, doesn't work with brokers other than Redis). As I've been using Celery I've accumulated a bunch of workarounds in my codebase and it's now peppered with comments like "TODO: find an alternative to Celery". Recently, I came across a bug and lost almost a day to debugging. This isn't the first time this has happened so I guess it was the final straw that led me to look into simpler alternatives.

As a general rule, I like to use as lightweight a library as possible for my needs and if possible one whose codebase I'm comfortable diving into. Celery just didn't fit this bill for me. I'm sure it's incredibly powerful but I just don't need that at this point—I just want something that I can easily understand, works when I expect it to, and where I don't dread modifying parts of my codebase that touch it.

**Comment by distefam** [comment]
In your example above, you say in a comment to "block until the four fetch tasks have returned". It seems to me that an easy way to do that would be to modify the following line of your example to be:

```python
results = [task_result.get(blocking=True) for task_result in tasks]
```

Since I intend to wait for all tasks to finish before returning results this seems like it would work. However, did you have something else in mind?

**Comment by coleifer** [comment]
Right, sorry I mistyped my example -- I had meant to add `blocking=True`. Updated comment as well.

**Comment by distefam** [comment]
Great, thank you again. I'm going to close this issue then.

**Comment by slidenerd** [comment]
Apologies for the newbie question, If you say blocking=True what happens when one of the subtasks fails and there is no retry or worker shuts down and restarts due to some issue

**Comment by coleifer** [comment]
If the task fails, all bets are off. The task will return upon failure and the `TaskResultWrapper` will have an empty result for the failed task.

**Comment by coleifer** [comment]
Committed: d6a5f372715737cbfbed6ff71dc685107ade1e5e

The above commit adds a `reset()` method to `TaskResultWrapper` so that you can re-get() the value of a task that failed and was subsequently retried successfully.

**Comment by AzMoo** [comment]
> I just read the celery docs for `group` and `chord`...my own opinion is that it's hideously complicated and hard to understand (littered with `.s()`, `.si()`, pipes, etc). I recognize the engineering talent that's required to make a library like celery, but if I was reviewing code that looked like those examples I'd definitely take issue with it on account of it being almost unreadable.
> 
> That being said, I think there's probably an analogous pattern in threading for spawning a bunch of threads (which may put values in a queue), then waiting for all to join and doing something with the values put in the queue.
> 
> To implement that kind of thing with huey you'll want to do something like:
> 
> ```python
> @huey.task()
> def fetch_from_api(resource):
>     return api.get(resource)
> 
> @huey.task()
> def initialize_data_from_api():
>     resources = ['foo', 'bar', 'baz', 'something']
> 
>     # Spawn 4 tasks.
>     tasks = [fetch_from_api(resource) for resource in resources]
> 
>     # Block this task until the 4 fetch tasks have returned.
>     results = [task_result.get(blocking=True) for task_result in tasks]
> 
>     # Insert the resulting data into the db.
>     insert_into_db(results)
> ```
> 
> In the above code you have a master task that spawns individual API-call tasks and then _blocks_ until all tasks have been executed and returned, before inserting the resulting data into the db. Obviously for the above to work you'll need at least 2 workers, since the master task is going to block up a worker for the duration of the API calls.
> 
> Is there any particular reason you're migrating away from celery?

Hi, I'm sorry for dredging up such an old issue but it seemed the most appropriate spot to talk about the problem I'm having. 

I'm attempting to implement this pattern as you've described but I'm running into trouble when running in pytest with `always_eager` turned on. 

As you're probably aware when a task is executed with always_eager enabled it executes immediately which is expected, but the issue is that it's returning the result of the task instead of a `TaskResultWrapper` instance. This means that I'm unable to call `task_result.get(blocking=True)` during unit and integration tests because the value in `tasks` is not a TaskResultWrapper instance and the call to `.get` fails.

I can work around the problem by checking if `always_eager` is enabled and skipping the `results` line, but it seems odd to me that a task returns a different type depending on if `always_eager` is enabled. I think it would be beneficial to return a `TaskResultWrapper` instance with the result being immediately available to provide a consistent return type regardless of whether `always_eager` is enabled.

**Comment by coleifer** [comment]
Correct. When running in `always_eager` mode, then the return value is whatever is returned by the task itself. If you are executing a pipeline in `always_eager` mode, then the pipeline is executed and a list is returned with all the result values.

I do understand your point, though, and I dislike that the return values are different. The `TaskResultWrapper` object exists to facilitate communication between the calling-code and an opaque "result store". The fact that `always_eager=True` should not really come into play.

Let me see about unifying these interfaces.

I am concerned, though, that this might break things for other people who are depending on the current behavior.

**Comment by coleifer** [comment]
You know...I don't like to say this, but I think it's just going to have to remain as-is. There's too much going on to try and achieve a 100% faithful replication of the ordinary "execute task via consumer" logic, without some refactoring. Error handling, for one, is a bit tricky.

I can suggest that if you need to unit-test some tasks, then test them as pure functions. Either by implementing a normal function and storing the task-decorated wrapper under a new name, or by using the decorated function's `.call_local()` helper.

**Comment by AzMoo** [comment]
Would you be open to a pull request that does some refactoring if I could work it out? I share the concern about breaking code that depends on this behaviour, but I feel the inconsistency is egregious enough to warrant a breaking change in a major version.

**Comment by coleifer** [comment]
Created #384

**Comment by jedie** [comment]
Oh: I also implement a context manager to use "always eager" in tests:


```
class HueyAlwaysEager:
    def __enter__(self):
        HUEY.always_eager = True
        return self

    def __exit__(self, *exc):
        HUEY.always_eager = False
```

e.g.:
```
    def test_decorator(self):
        self.assertFalse(HUEY.always_eager)

        with HueyAlwaysEager():
            self.assertTrue(HUEY.always_eager)

        self.assertFalse(HUEY.always_eager)
```

**Comment by henribru** [comment]
> I'm looking into adding support for chord-like API, but don't want to commit to an implementation as I'm not sure how cleanly I can do it.

Any news on this? Personally I'm not a huge fan of the workaround of calling `get` on each result since it blocks a worker. If you want to nest such tasks it's even worse since you could block many workers

**Comment by coleifer** [comment]
https://huey.readthedocs.io/en/latest/guide.html#task-pipelines

**Comment by henribru** [comment]
Maybe I'm missing something but can the current pipeline functionality be used if you'd like to wait for a group of tasks running in parallel?

**Comment by coleifer** [comment]
So you want to start a bunch of tasks and then when they're all done - do something with the results. In this case you can use `.map()` or just spawn them all and collect the result handles in a list (or use `ResultGroup`). These approaches imply calling `.get()` on the result handles. If you want to attach a callback when a certain task is finished, you can use `.then()`. There is no higher-level abstraction for a collection of parallel-running tasks.

**Comment by ppawlak** [comment]
Hey, I just found this issue for the same reason as others: do something when a group of tasks running in parallel are all done.

[This code snippet](https://github.com/coleifer/huey/issues/286#issuecomment-356818549) from 3 years ago works great **as long as the tasks returns something**. In my case they don't and `task_result.get(blocking=True)` just hangs forever. 
Is it on purpose? 3 years later is there any simple way to know if a task is done, like a flag or a method call on the `TaskResultWrapper `?

Adapting from the above snippet my use case would be something like this:
```python
@huey.task()
def push_to_api(resource):
    api.post(resource)
    # don't return anything

@huey.task()
def initialize_data_to_api():
    resources = ['foo', 'bar', 'baz', 'something']

    # Spawn 4 tasks.
    tasks = [push_to_api(resource) for resource in resources]

    # Block this task until the 4 push tasks have returned.
    results = [task_result.get(blocking=True) for task_result in tasks]

    # do something
    do_something()
```

**Comment by coleifer** [comment]
You can set the result store to store "None" values, but this can quickly fill-up your result storage if you are not reading and resolving those values (resolving a result is a destructive read). So it is good practice that **if you need to block while waiting on a result, to have the task return a non-None value**.

**Comment by ppawlak** [comment]
Thanks for the explanation @coleifer 

> if you need to block while waiting on a result, to have the task return a non-None value.

I'll stick to this solution then
